### PR TITLE
MastClass query helper function

### PR DESCRIPTION
**Pull request by mstcyr2**
The MAST API requires users to use JSON/Python Dictionaries to query MAST services. We want to create a helper function to format parameter dictionaries to make `astroquery.mast.Mast` queries easier for the users.

### CODE REVIEW

File: docs/mast/mast_mastquery.rst [hunk 1]
```diff
@@ -3,26 +3,118 @@
...
-
 The Mast class provides more direct access to the MAST interface.  It requires
 more knowledge of the inner workings of the MAST API, and should be rarely
 needed.  However in the case of new functionality not yet implemented in
 astroquery, this class does allow access.  See the `MAST api documentation
-<https://mast.stsci.edu/api>`_ for more information.
+<https://mast.stsci.edu/api>`__ for more information.
+
+The basic MAST query function accepts
+`MAST Service <https://mast.stsci.edu/api/v0/_services.html>`__-specific
```
Comment (jaymedina) [inline docs/mast/mast_mastquery.rst:13]: I would reword this slightly to: _The basic MAST query function allows users to query through the following `MAST Services <https://mast.stsci.edu/api/v0/_services.html>`__ using their corresponding parameters_ (with the exception of the service that you mentioned `mast_query` doesnt account for).

File: docs/mast/mast_mastquery.rst [hunk 2]
```diff
@@ -3,26 +3,118 @@
...
+parameters and returns query results as an `~astropy.table.Table`.
+
+Filtered Mast Queries
+=====================
+
+MAST's Filtered services use the parameters 'columns' and 'filters'. The 'columns'
+parameter is a required string that specifies the columns to be returned as a
+comma-separated list. The 'filters' parameter is a required list of filters to be
+applied. The `~astroquery.mast.MastClass.mast_query` method accepts that list of
```
Comment (jaymedina) [inline docs/mast/mast_mastquery.rst:22]: Be sure to mention that the criteria under the `filters` parameter work just the same way as the criteria used in `Observations.query_criteria` for example. Would be good to draw parallels to `ObservationsClass` as its familiar to our users.

File: docs/mast/mast_mastquery.rst [hunk 3]
```diff
@@ -3,26 +3,118 @@
...
+filters as keyword arguments paired with a list of values.
+
+.. doctest-remote-data::
+
+   >>> from astroquery.mast import Mast
+   ...
+   >>> observations = Mast.mast_query('Mast.Jwst.Filtered.Nirspec',
+   ...                                columns = 'title, instrume, targname',
```
Comment (jaymedina) [inline docs/mast/mast_mastquery.rst:30]: Where can people go to see what options they have for `filters` and `columns`? i.e. How would a user know that the column name is `instrume` and not `instrument`? Be sure to post a resource pointing to this.

File: docs/mast/mast_mastquery.rst [hunk 4]
```diff
@@ -3,26 +3,118 @@
...
+   Traceback (most recent call last):
+   ...
+   astroquery.exceptions.RemoteServiceError: Incorrect syntax near '*'.
+
+
+Cone Searches
+=============
+
+MAST's cone search services use the parameters 'ra', 'dec', and 'radius' and return
+a table of observations with all columns present.
```
Comment (jaymedina) [inline docs/mast/mast_mastquery.rst:108]: Is this the service that ignores when the user specify which `columns` to return? If so, be sure to mention this somewhere in here before the example. Is there error handling to catch when this happens?

File: docs/mast/mast_mastquery.rst [hunk 5]
```diff
@@ -51,28 +143,43 @@ The basic MAST query function returns query results as an `~astropy.table.Table`
...
 
 .. doctest-remote-data::
 
    >>> from astroquery.mast import Mast
    ...
-   >>> service = 'Mast.Caom.Cone'
-   >>> params = {'columns': "*",
-   ...           'filters': {}}
-   >>> observations = Mast.service_request(service, params)
+   >>> observations = Mast.mast_query('Mast.Caom.Cone', columns = 'ra')
```
Comment (jaymedina) [inline docs/mast/mast_mastquery.rst:155]: I'm a little confused about this one, why was columns changes to 'ra' from '*'? This example is meant to show how not using the right filters leads to a `TraceBackError`, so what about adding a non-coordinate filter to the cone search, since cone search expects RA/Dec and not Teff, for example?

File: docs/mast/mast_mastquery.rst (hunk 4 above)
Comment (jaymedina) [reply docs/mast/mast_mastquery.rst:108]: Nvm! I see that you describe this a few examples below. You can ignore this comment

File: docs/mast/mast_mastquery.rst [hunk 6]
```diff
@@ -51,28 +143,43 @@ The basic MAST query function returns query results as an `~astropy.table.Table`
...
+
+Advanced Service Request
+========================
 
-If the output is not the MAST json result type it cannot be properly parsed into a `~astropy.table.Table`.
-In this case, the async method should be used to get the raw http response, which can then be manually parsed.
+Certain MAST Services, such as `Mast.Name.Lookup
+<https://mast.stsci.edu/api/v0/_services.html#MastNameLookup>`__ will not work with
+`astroquery.mast.MastClass.mast_query` due to it's return type. If the output of a query
+is not the MAST json result type it cannot be properly parsed into a`~astropy.table.Table`.
```
Comment (jaymedina) [inline docs/mast/mast_mastquery.rst:180]: There's a typo here, the a before `~astropy.table.Table`

File: astroquery/mast/observations.py [hunk 7]
```diff
@@ -908,6 +908,42 @@ def service_request_async(self, service, params, *, pagesize=None, page=None, **
...
         return self._portal_api_connection.service_request_async(service, params, pagesize, page, **kwargs)
 
+    def mast_query(self, service, columns='*', **kwargs):
+        # Specific keywords related to positional and MashupRequest parameters.
+        position_keys = ['ra', 'dec', 'radius', 'position']
+        request_keys = ['format', 'data', 'filename', 'timeout', 'clearcache',
+                        'removecache', 'removenullcolumns', 'page', 'pagesize']
+
+        # Explicit formatting for Mast's filtered services
+        if 'Filtered' in service:
```
Comment (jaymedina) [inline astroquery/mast/observations.py:918]: Currently it seems like the input `service` argument is case-sensitive, and I think people will have trouble with that if they don't capitalize the right letters. Can you make the `service` argument case-insensitive?

**Review by jaymedina — CHANGES_REQUESTED** [review]
Looking good so far, thanks for this work! Let me know if you have any questions about my requested changes and/or comments.

File: docs/mast/mast_mastquery.rst (hunk 5 above)
Comment (mstcyr2) [reply docs/mast/mast_mastquery.rst:155]: Columns value is defaulted to '*' when not specified so I just changed it to something users might be looking for in a cone search. And I see what you mean about the other filters, I will add that.

**Review by jaymedina — APPROVED** [review]
This looks good to me! I will defer to @bsipocz for any other comments.

File: astroquery/mast/observations.py [hunk 8]
```diff
@@ -908,6 +908,45 @@ def service_request_async(self, service, params, *, pagesize=None, page=None, **
 
         return self._portal_api_connection.service_request_async(service, params, pagesize, page, **kwargs)
 
+    def mast_query(self, service, columns=None, **kwargs):
```
Comment (bsipocz) [inline astroquery/mast/observations.py:911]: we're trying to get rid of the ``**kwargs`` in all signatures, at least where it's not a convenience method falling back on totally different signatures (think about `Table.read()`), and where the list of potential kwargs are not unmanagably long. 

So, if possible, please spell them out.

Comment (bsipocz) [reply astroquery/mast/observations.py:911]: Also, please make optional agruments kwarg only. And please add documentation as docstring

**Review by bsipocz — CHANGES_REQUESTED** [review]
I have two things before we can merge this:

- docstrings are needed for any public functionality
- please avoid using the `**kwargs` syntax in a signature as it masks over potential issues, keywords the user think have any effect while they are ignored, etc. We do this type of cleanup slowly all over the library, so no new things should add one, unless there is a very convincing argument along with it.

File: astroquery/mast/observations.py (hunk 8 above)
Comment (mstcyr2) [reply astroquery/mast/observations.py:911]: I think `**kwargs` are necessary here because `mast_query` deals with requests to all the different mashup services and uses kwargs for each service's list of properties. I also addressed the mashup property parameters using kwargs because they're all optional for the `Mast.service_request` call in the return statement. The api also does error-handling for properties or kwargs that don't exist within a service.

Comment (jaymedina) [reply astroquery/mast/observations.py:911]: Hi @bsipocz , Makayla is right. The arguments here depend on which database is being accessed, so we need `**kwargs` as a catchall. What do you mean by adding documentation as docstring? In this PR Makayla added a link to the MAST documentation a few lines below (L924 currently).

**Comment by jaymedina** [comment]
Just a heads up, I'll be taking over this PR for Makayla since she completed her internship ( 🎉 )

**Comment by bsipocz** [comment]
Congrats @mstcyr2 on finishing your internship. And thank you for all the contributions to astroquery.

@jaymedina - could you have a look at the remote tests, e.g. I see failures mast_mastquery.rst.

**Comment by jaymedina** [comment]
> @jaymedina - could you have a look at the remote tests, e.g. I see failures mast_mastquery.rst.

Thanks for looking into this. Is it that you saw doc build failures in `mast_mastquery.rst`? I just built the docs locally with no errors and then ran each of the examples manually in a notebook and all seems to work fine. My latest commit fixes syntax in the examples by removing the spaces around `=` in the kwarg calls, which I suspect might have led to what you were seeing? I also just ran the `test_mast_remote.py` unit tests and there are no failures related to the MAST query helper function. Let me know if there is something else I should check. 

Thanks!

**Comment by bsipocz** [comment]
> Thanks for looking into this. Is it that you saw doc build failures caused by mast_mastquery.rst

No, I see failures when run the doctests in that file.

``pytest docs/mast -R``

**Comment by bsipocz** [comment]
This needs a rebase, I tried to use the "newsish" github conflict resolution with a rebase, but it still only was able to do the merge approach.

**Comment by jaymedina** [comment]
Just did a rebase and squash @bsipocz let me know if there's anything else I should update. Thanks!

**Review by bsipocz — APPROVED** [review]
